virtcontainers/qemu.go

```diff
@@ -703,6 +704,10 @@ func (q *qemu) addDeviceToBridge(ID string) (string, Bridge, error) {
 	var err error
 	var addr uint32
 
+	if len(l.state.Bridges) == 0 {
+		return "", Bridge{}, errors.New("failed to get available address from bridges")
```

Comment:
Well this format is not appropriate as you return a message that is not true about the real error here. There is no bridge, so obviously there is no address available.
What I'm trying to say is that your error message is valuable in my opinion to be added to every error to this function, except that you need to add a specific message when there's no bridge.

Reply:
I was going to agree here (return a `nil` error here and return a real error after the `for` loop below), but this is more subtle than it first appears... 

- `qemu.init()` calls `hypervisorConfig.valid()` and that function does this:
  ```golang
  if conf.DefaultBridges == 0 {
      conf.DefaultBridges = defaultBridges     
  }
  ```
- Then `qemu.init()` calls:
  ```golang
  q.state.Bridges = q.arch.bridges(q.config.DefaultBridges)
  ```

So, `len(q.state.Bridges) == 0` is currently an error scenario. However, I agree that we need another error at the end of this function in case there are no free bridge slots.

@Ace-Tang - this is a rather interesting PR! What it also shows me is that we don't have any unit tests for `QemuState`. It would be great if you could add a simple unit test for the three scenarios we're talking about here:

- `len(q.state.Bridges == 0)`
- no more bridge slots available
- successful call to `addDeviceToBridge()`
- (and ideally a test for `removeDeviceFromBridge()`).